File: 2023/P08/problem8a.rs
use std::fs;
use std::collections::HashMap;

fn main() {
    let data = fs::read_to_string("problem8.txt")
        .expect("Failed to read file");
    let data = data
        .trim()
        .split('\n')
        .collect::<Vec<_>>();
    let insts = data[0];
    let mut data_map = HashMap::<String, (String, String)>::new();
    let data = &data[2..];
    for line in data {
        let parts = line
            .split(" = ")
            .collect::<Vec<_>>();
        let key = parts[0];
        let l = parts[1].len();
        let l_r_vec = parts[1][1..l - 1]
            .split(", ")
            .collect::<Vec<_>>();
        let l = l_r_vec[0];
        let r = l_r_vec[1];
        data_map.insert((*key).to_string(), ((*l).to_string(), (*r).to_string()));
    }
    let mut curr_char = insts.chars();
    let mut curr = "AAA";
    let mut steps = 0;
    while curr != "ZZZ" {
        let pair = data_map.get(curr);
        match pair {
            Some(s) => {
                match curr_char.next() {
                    Some(c) => { 
                        match c {
                            'L' => { curr = &s.0; },
                            'R' => { curr = &s.1; },
                            _ => { println!("Error: {c} is not L or R"); }
                        }
                    },
                    None => {
                        curr_char = insts.chars();
                        continue;
                    }
                }
                steps += 1;
            },
            None => {
                println!("Error: {curr} not found in data_map");
            }
        }
    }
    println!("{steps}");
}

File: 2024/P08/problem8a.py
from collections import defaultdict

data = []
with open("problem8.txt", 'r') as f:
    data = f.read().strip('\n').split('\n')

antinodes = set()
antennae = defaultdict(set)
for i, line in enumerate(data):
    for j, c in enumerate(line):
        if c != '.':
            antennae[c].add((i, j))

for c, points in antennae.items():
    for p1 in points:
        for p2 in points:
            if p1 == p2:
                continue

            dx, dy = (p2[0] - p1[0], p2[1] - p1[1])
            p3 = (p2[0] + dx, p2[1] + dy)
            p4 = (p1[0] - dx, p1[1] - dy)
            if 0 <= p3[0] < len(data) and 0 <= p3[1] < len(data[0]):
                antinodes.add(p3)
            if 0 <= p4[0] < len(data) and 0 <= p4[1] < len(data[0]):
                antinodes.add(p4)

print(len(antinodes))
